src/globus_sdk/scopes/scope_definition.py

```diff
@@ -236,6 +236,61 @@ def __repr__(self) -> str:
     def __str__(self) -> str:
         return self.serialize()
 
+    def __contains__(self, other: t.Any) -> bool:
+        """
+        scope1 in scope2 is defined as subtree matching.
+
+        A scope contains another scope if
+        - the top level strings match
+        - the optional-ness matches OR only the contained scope is optional
+        - the dependencies of the contained scope are all contained in dependencies of
+          the containing scope
+
+        Therefore, the following are true:
+
+        .. code-block:: pycon
+
+            # self inclusion works
+            >>> Scope.deserialize("foo") in Scope.deserialize("foo")
+            # optional mismatches in either direction do not indicate containment
+            >>> Scope.deserialize("foo") not in Scope.deserialize("*foo")
+            >>> Scope.deserialize("*foo") not in Scope.deserialize("foo")
+            # dependencies have the expected meanings
+            >>> Scope.deserialize("foo") in Scope.deserialize("foo[bar]")
+            >>> Scope.deserialize("foo[bar]") not in Scope.deserialize("foo")
+            >>> Scope.deserialize("foo[bar]") in Scope.deserialize("foo[bar[baz]]")
+            # dependencies are not transitive and obey "optionalness" matching
+            >>> Scope.deserialize("foo[bar]") not in Scope.deserialize("foo[fizz[bar]]")
+            >>> Scope.deserialize("foo[bar]") not in Scope.deserialize("foo[*bar]")
+        """
+        # scopes can only contain other scopes
+        if not isinstance(other, Scope):
+            return False
+
+        # top-level scope must match
+        if self._scope_string != other._scope_string:
+            return False
+        # if both are optional, okay
+        # if neither is optional, okay
+        # but if only one is optional...
+        if self.optional != other.optional:
+            # ... then make sure it is 'other'
+            if self.optional:
+                return False
```

Comment:
I had to stare at this for a moment, even with the comments. This may have been that the final explanatory comment mentions `other` but the actual condition was for `self`. Would this -- still with explanatory comments! -- improve the readability here?

```python
# Three out of the four possible combinations are acceptable:
# Both are optional; neither is optional; or only 'other' is optional.
# If 'other' is required but 'self' is optional,
# then 'self' cannot contain 'other'.
if not other.optional and self.optional:
    return False
```

Reply:
Ooh, I like how this flattens the conditions. It is written the way I _reasoned_ through what is correct. But now, with the conclusion in hand, I like it better with the explanation in the comment but the condition being this way around. It makes it clear that there's only one False case, and calls out what that is.
I almost want to put a truth table into the comment! 😁

Reply:
I'll rubber-stamp any change that adds a truth table, but only if the matrix intersection points are Unicode characters instead of "T/F" or "Y/N" or similar. 😆